core/blockchain_snapshot_test.go

```diff
@@ -0,0 +1,469 @@
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken snapshot.
+// In this case the chain head should be rewound to the point
+// with available state. And also the new head should must be lower
+// than disk layer. But there is no committed point so the chain
+// should be rewound to genesis.
+func TestNoCommitCrashWithNewSnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             false,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        0,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       0,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken snapshot.
+// In this case the chain head should be rewound to the point
+// with available state. And also the new head should must be lower
+// than disk layer. But there is only a low committed point so the
+// chain should be rewound to committed point.
+func TestLowCommitCrashWithNewSnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             false,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        2,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       2,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken snapshot.
+// In this case the chain head should be rewound to the point
+// with available state. And also the new head should must be lower
+// than disk layer. But there is only a high committed point so the
+// chain should be rewound to genesis.
+func TestHighCommitCrashWithNewSnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             false,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        6,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       0,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken and "legacy format"
+// snapshot. In this case the entire legacy snapshot should be discared
+// and rebuild from the new chain head.
+func TestNoCommitCrashWithLegacySnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             true,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        0,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       0,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
```

Comment:
In the comment you say that "the entire legacy snapshot should be discarded and rebuild from the new chain head", yet the tests says the snapshot gets recovered at block 4. 

Either the comment is wrong or the test (that passes) is. Please clarify.

Reply:
This case should correspond to:

```
	// Chain:
	//   G->C1->C2->C3->C4->C5->C6->C7->C8 (HEAD)
	//
	// Commit:   G
	// Snapshot: G, C4
	//
	// CRASH
	//
	// ------------------------------
	//
	// Expected in leveldb:
	//   G->C1->C2->C3->C4->C5->C6->C7->C8
	// Expected head header    : C8
	// Expected head fast block: C8
	// Expected head block     : G
	// Expected snapshot disk  : C4
```